File: client/site_tests/platform_MiniJailPtraceDisabled/platform_MiniJailPtraceDisabled.py
# Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import logging
import os
import re

from autotest_lib.client.bin import test
from autotest_lib.client.common_lib import error, utils

class platform_MiniJailPtraceDisabled(test.test):
    version = 1
    preserve_srcdir = True

    def setup(self):
        os.chdir(self.srcdir)
        utils.system('make clean')
        utils.system('make all')


    def __run_cmd(self, cmd):
        result = utils.system_output(cmd, retain_output=True,
                                     ignore_status=True)
        return result


    def run_once(self):
        # Check that --disable-tracing works
        check_cmd = os.path.join(self.bindir, 'platform_MiniJailPtraceDisabled')
        cmd = ('/sbin/minijail --disable-tracing -- ' + check_cmd)
        result = self.__run_cmd(cmd)
        succeed_pattern = re.compile(r"SUCCEED: (.+)")
        success = succeed_pattern.findall(result)
        if len(success) == 0:
          raise error.TestFail('Tracing of a child process was allowed')

File: client/site_tests/platform_MiniJailPtraceDisabled/src/platform_MiniJailPtraceDisabled.cc
// Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <sys/ptrace.h>
#include <sys/types.h>
#include <unistd.h>

// Calls fork() and sees if it can do a PTRACE_ATTACH on the parent.  A properly
// jailed process should not be ptrace-able by another process.
void CheckPtraceDisabled() {
  pid_t parent_pid = getpid();
  if (fork() == 0) {
    if (ptrace(PTRACE_ATTACH, parent_pid, NULL, NULL) != -1) {
      printf("FAIL: ptrace attach of %d succeded.\n", parent_pid);
      ptrace(PTRACE_DETACH, parent_pid, NULL, NULL);
    } else {
      printf("SUCCEED: ptrace attach of %d failed (errno %d).\n", parent_pid,
             errno);
    }
    kill(parent_pid, SIGKILL);
  } else {
    sleep(10000);
  }
}

int main(int argc, const char* argv[]) {
  CheckPtraceDisabled();

  return 0;
}
